lib/util/buffer-manager.ts

```diff
@@ -0,0 +1,37 @@
+/// <reference path='../../typings/tsd.d.ts' />
+
+export interface IBufferedData<T> {
+    data: T[];
+    overflow: number;
+}
+
+export class BufferManager<T> {
+
+    // maxBuffLength: Maximum numbers of data item per level of buffer
+    // maxHistoryDepth: Maximum numbers of preserved buffer history
+    constructor(maxBuffLength: number, maxHistoryDepth: number) {
+        // Not Implemented Yet
+    }
+
+    // Push one data item to the current buffer
+    push(data: T): void {
+        // Not Implemented Yet
+        throw new Error();
+    }
+
+    // Get buffered data array.
+    // depth: This optional parameter controls which level of history will be retrieved
+    // depth: Default is to retrieve the current buffer. The first history will be depth 1
+    // Return an object containing the data array and an integer for number of overlow items
+    get(depth?: number): IBufferedData<T> {
+        // Not Implemented Yet
+        throw new Error();
+    }
+
+    // Flush the last data array from the history and move the entire history one step forward
+    // Subsequent push will be stored into a new array
+    flush(): void {
+        // Not Implemented Yet
+        throw new Error();
+    }
+}
```

Comment:
@ericvw @yupinghu As discussed yesterday, I added this `BufferManager` class skeleton for review purpose. Once we settled on the interfaces, I can go ahead and implement it.